### ISSUE TITLE
Controller-less Ahoy tracking is not working

### ISSUE THREAD

**Issue by Startouf** [ankane/ahoy#410]
I am opening an issue related to the [comment here](https://github.com/ankane/ahoy/issues/338#issuecomment-371469445)

The current code of ahoy v2 and v3 do not support tracking without controller, as the tracking code is coupled to a controller's request and uses extensively `controller.request` the request. 

```ruby
::Ahoy::Tracker.new(controller: nil).track(..)
```

Will cause several errors like

> undefined method `cookie_jar' for nil:NilClass

> undefined method `env' for nil:NilClass

The problem arises because of [lines like this one](https://github.com/ankane/ahoy/blob/1fbdca705263122222fd72c3e376df816652ee7b/lib/ahoy/tracker.rb#L179) that assume a request is always present

Use cases for using Ahoy without controller have been mentioned in the linked issue, one of them being handling user actions done outside a controller scope, such as when the website has implemented a "reply by email" that does not involve any controller at all, but just workers.

It would be possible to add `.try` everywhere as in `request.try { |r| r.something }` but that would make the code a bit ugly. A cleaner approach would involve using some strategy pattern to select an implementation of tracking methods entirely decoupled from a controller's request, or to provide a mocked request environment that would fit asynchronous scenarios.

Furthermore, it would be nice to add an attr_reader on the tracker class for the `@user` variable.

Also, since the need arise to instanciate an Ahoy tracker from a background job, in order to attach the user to the event (for example: I have a "reply by email" service working via AWS services + queues, so there is indeed a user action but no controller involved)

Right now I need to `ahoy.instance_variable_set(:@user, user)` to set manually the user in my service tracker.

**Comment by ankane** [comment]
Hey @Startouf, Ahoy isn’t designed to be used outside of a request, as events require a visit and there’s no concept of a visit outside a request (you could pass in visit and visitor tokens manually I suppose). If you want to track events without visits, you’ll need to modify the model code to make visits optional, then write directly to the events table.

**Comment by Startouf** [comment]
So far I've been using ahoy tracking even without controller/visits on v1, it hasn't been a problem until I tried to update. And it seems it is working fine when I checkout my fork with patch #411. I have application services that that take a controller as optional parameter and the following implementation has been working perfectly so far in v1, so I'm just trying to make it work as well in v3

```
class MyService
  def initialize(controller: nil)
    @controller = controller
  end

  def ahoy
    @ahoy ||= controller.present? ? controller.ahoy : service_ahoy
  end

  def service_ahoy
    ::Ahoy::Tracker.new(controller: nil)
  end
```

**Comment by Startouf** [comment]
> Ahoy isn’t designed to be used outside of a request

I understand this but then this makes it unusable for us in practice with our use cases unless we make it compatible without request. If you decide not to support this this means I'll have to stay on our own fork of the project. Not necessarily a bad deal for us, we'd probably take the opportunity to integrate Ahoy code directly in our project and keep only what we're using, but it only takes a couple edits to the code (cf attached PR) and the gem becomes useable without controller/request as it was in v1.

**Comment by ankane** [comment]
Addressed by #411
